File: level2/다리를_지나는_트럭.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int total_weight = 0;
    vector<int> truck_on_bridge;
    int bridge[bridge_length];
    int temp;

    for(int i=0; i<bridge_length; i++)
        bridge[i] = 0;

    while(truck_weights.size() != 0){

        if(bridge[bridge_length-1] != 0){
            total_weight -= truck_on_bridge[0];
            truck_on_bridge.erase(truck_on_bridge.begin());
        }

        for(int i=bridge_length-1; i>0; i--)
            bridge[i] = bridge[i-1];

        bridge[0] = 0;

        if(total_weight + truck_weights[0] <= weight){
            total_weight += truck_weights[0];
            truck_on_bridge.push_back(truck_weights[0]);
            bridge[0] = truck_weights[0];
            truck_weights.erase(truck_weights.begin());
        }
        answer++;
    }

    int index = bridge_length;

    for(int i=0; i<bridge_length; i++)
        if(bridge[i] != 0 ){
            index = i;
            break;
        }
        
    return answer+(bridge_length-index);
}

File: JavaScript/level2/다리를_지나는_트럭.js
function solution(bridge_length, weight, truck_weights) {
    let answer = 0;
    let total_weight = 0;
    let turck_on_bridge = [];
    let time_to_stay = [];
    
    while(!(turck_on_bridge.length === 0 && truck_weights.length === 0)){
        if(truck_weights.length !== 0)
            if(truck_weights[0]+total_weight <= weight){
                total_weight += truck_weights[0];
                turck_on_bridge.push(truck_weights[0]);
                time_to_stay.push(bridge_length);
                truck_weights.shift();
            }

        time_to_stay = time_to_stay.map((i) => i-1);
  
        if(time_to_stay[0] === 0){
            time_to_stay.shift();
            total_weight -= turck_on_bridge[0];
            turck_on_bridge.shift();
        }
        answer++;
    }
    
    return answer+1;
}